src/lib.rs

```diff
@@ -1,91 +1,62 @@
-use cfgrammar::RIdx;
-use lrlex::{lrlex_mod, DefaultLexeme, DefaultLexerTypes, LRNonStreamingLexerDef};
-use lrpar::{lrpar_mod, Lexeme, Node};
+use lrlex::{lrlex_mod, DefaultLexerTypes};
+use lrpar::{lrpar_mod, LexParseError, NonStreamingLexer};
 
 lrlex_mod!("calc.l");
 lrpar_mod!("calc.y");
 
-pub fn eval(
-    lexerdef: &LRNonStreamingLexerDef<DefaultLexerTypes>,
-    input: &str,
-) -> Result<String, String> {
-    let lexer = lexerdef.lexer(input);
-    let (res, errs) = calc_y::parse(&lexer);
-    for e in errs {
-        println!("{}", e.pp(&lexer, &calc_y::token_epp));
-    }
-    match res {
-        Some(x) => {
-            return Ok(format!("{}", Eval::new(input).eval(&x)));
-        }
-        _ => Err(String::from("Unable to evaluate expression.")),
-    }
-}
+mod ast;
 
-struct Eval<'a> {
-    s: &'a str,
-}
+use ast::Opcode;
 
-// stolen from: https://github.dev/softdevteam/grmtools
-impl<'a> Eval<'a> {
-    fn new(s: &'a str) -> Self {
-        return Eval { s };
-    }
+pub struct Calc {}
+
+impl Calc {
+    pub fn from_str(input: &str) -> Result<Vec<Opcode>, String> {
+        let lexer_def = calc_l::lexerdef();
+        let lexer = lexer_def.lexer(input);
+        let (ast_exp, errs) = calc_y::parse(&lexer);
 
-    fn eval_r_exp(&self, nodes: &Vec<Node<DefaultLexeme, u32>>) -> i64 {
-        if nodes.len() == 1 {
-            return self.eval(&nodes[0]);
-        } else {
-            debug_assert_eq!(nodes.len(), 3);
-            return self.eval(&nodes[0]) + self.eval(&nodes[2]);
+        let err_msg = Self::get_parse_err(&lexer, errs);
+        if err_msg.is_empty() == false {
+            return Err(err_msg);
         }
-    }
 
-    fn eval_r_term(&self, nodes: &Vec<Node<DefaultLexeme, u32>>) -> i64 {
-        if nodes.len() == 1 {
-            return self.eval(&nodes[0]);
-        } else {
-            debug_assert_eq!(nodes.len(), 3);
-            return self.eval(&nodes[0]) * self.eval(&nodes[2]);
+        match ast_exp {
+            Some(res) => match res {
+                Ok(exp) => Ok(exp),
+                Err(_) => Err(err_msg),
+            },
+            None => Err(err_msg),
         }
     }
 
-    fn eval_r_factor(&self, nodes: &Vec<Node<DefaultLexeme, u32>>) -> i64 {
-        if nodes.len() == 1 {
-            if let Node::Term { lexeme } = nodes[0] {
-                self.s[lexeme.span().start()..lexeme.span().end()]
-                    .parse()
-                    .unwrap()
-            } else {
-                unreachable!();
-            }
-        } else {
-            debug_assert_eq!(nodes.len(), 3);
-            self.eval(&nodes[1])
+    pub fn eval(opcodes: Vec<Opcode>) -> Result<u64, String> {
+        for opcode in opcodes {
+            return Self::eval_exp(opcode);
```

Comment:
I find this loop weird cause I'm returning on the first iteration. From what I've seen I always got a single opcode but I was asked to implement `Vec<Opcode>`

Reply:
Oh, I missed this bit.

At this stage, where we are evaluating an AST, you'd just evaluate the root of the tree, right?

(I spoke to laurie yesterday, and the next stage will be to reduce your AST to a vector of bytecodes. Perhaps the confusion lies there?)

Reply:
> At this stage, where we are evaluating an AST, you'd just evaluate the root of the tree, right?
it starts from the root but it evaluates LHS/RHS recursively via `eval_exp`

Reply:
maybe It will be clearer to me when we'll move to the next stage - reducing the AST to a vector of bytecodes...

Reply:
> I think it starts from the root but it evaluates LHS/RHS recursively via eval_exp

that's right.

> maybe It will be clearer to me when we'll move to the next stage

ok!